src/node_locks.cc

```diff
@@ -0,0 +1,738 @@
+    }
+
+    // Call user callback
+    Local<Object> lock_info_obj =
+        CreateLockInfoObject(isolate,
+                             context,
+                             grantable_request->name(),
+                             grantable_request->mode(),
+                             grantable_request->client_id());
+    if (lock_info_obj.IsEmpty()) {
+      return;
+    }
+    Local<Value> callback_arg = lock_info_obj;
+    Local<Value> callback_result;
+    {
+      TryCatchScope try_catch_scope(env);
+      if (!grantable_request->callback()
+               ->Call(context, Undefined(isolate), 1, &callback_arg)
+               .ToLocal(&callback_result)) {
+        grantable_request->waiting_promise()
+            ->Reject(context, try_catch_scope.Exception())
+            .Check();
+        grantable_request->released_promise()
+            ->Reject(context, try_catch_scope.Exception())
+            .Check();
+        return;
+      }
+    }
+
+    // Allocate a LockHolder on the heap to safely manage the lock's lifetime
+    // until the user's callback promise settles.
+    auto* fulfill_holder = new LockHolder(granted_lock);
+    auto* reject_holder = new LockHolder(granted_lock);
+    Local<Function> on_fulfilled_callback;
+    Local<Function> on_rejected_callback;
+
+    if (!Function::New(context,
+                       OnLockCallbackFulfilled,
+                       External::New(isolate, fulfill_holder))
+             .ToLocal(&on_fulfilled_callback) ||
+        !Function::New(context,
+                       OnLockCallbackRejected,
+                       External::New(isolate, reject_holder))
+             .ToLocal(&on_rejected_callback)) {
+      delete fulfill_holder;
+      delete reject_holder;
+      return;
+    }
+
+    // Handle promise chain
+    if (callback_result->IsPromise()) {
+      Local<Promise> promise = callback_result.As<Promise>();
+      if (promise->State() == Promise::kRejected) {
+        Local<Value> rejection_value = promise->Result();
+        grantable_request->waiting_promise()
+            ->Reject(context, rejection_value)
+            .Check();
+        grantable_request->released_promise()
+            ->Reject(context, rejection_value)
+            .Check();
+        delete fulfill_holder;
+        delete reject_holder;
+        {
+          Mutex::ScopedLock scoped_lock(mutex_);
+          ReleaseLock(granted_lock.get());
+        }
+        ProcessQueue(env);
+        return;
+      } else {
+        grantable_request->waiting_promise()
+            ->Resolve(context, callback_result)
+            .Check();
+        USE(promise->Then(
+            context, on_fulfilled_callback, on_rejected_callback));
+      }
+    } else {
+      grantable_request->waiting_promise()
+          ->Resolve(context, callback_result)
+          .Check();
+      Local<Value> promise_args[] = {callback_result};
+      USE(on_fulfilled_callback->Call(
```

Comment:
Likewise here. Calling the callback could throw. The error should be appropriately propagated.